### PR TITLE
Breadth first search

### PR DESCRIPTION
**Pull request by RoiEXLab**
## Overview
Here comes the long awaited cleanup PR regarding the weird breadth-first-search 🎉 

## Functional Changes
None

## Manual Testing Performed
Verified the route looks right when playing

## Additional Review Notes
In case you're wondering: This is the graph I'm testing, it's hand drawn with the touchpad of my laptop so ...
Graph
Z is the destination (index 8) (Z is Ziel in german ^^) and S the start (index 0).

### CODE REVIEW

**Comment by RoiEXLab** [comment]
I'm in a bit of a hurry right now, so if the build is failing, I will fix it later this evening...
Also I tried to figure out why route updates are sometimes so slow.
I haven't figured out where the bottleneck is, though I can confirm it's not this algorithm!

File: game-core/src/test/java/games/strategy/engine/data/RouteFinderTest.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class RouteFinderTest {
+
+  private final PlayerID player = mock(PlayerID.class);
+  private final GameMap map = mock(GameMap.class);
+  private final List<Territory> territories = new ArrayList<>();
+
+  private final int[][] graph = {
```
Comment (ssoloff) [inline game-core/src/test/java/games/strategy/engine/data/RouteFinderTest.java]: Would an ASCII-art rendition of the graph you drew in the description be helpful for a future code reader to understand this matrix?

File: game-core/src/test/java/games/strategy/engine/data/RouteFinderTest.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class RouteFinderTest {
```
Comment (ssoloff) [inline game-core/src/test/java/games/strategy/engine/data/RouteFinderTest.java]: The test coverage here looks pretty good for just a single test.  The only suggestions I'd add are for some boundary conditions:

* When `start` and `end` are the same.
* When no route exists between `start` and `end`.

**Comment by ron-murhammer** [comment]
@RoiEXLab I'd really like to see some performance testing around this as that is extremely important for route finding. I'd recommend using Red Sun Over China as it has lots of land canals which makes checking each step a bit slower.

File: game-core/src/test/java/games/strategy/engine/data/RouteFinderTest.java (hunk 1 above)
Comment (RoiEXLab) [reply game-core/src/test/java/games/strategy/engine/data/RouteFinderTest.java]: Yes, however I wasn't able to find a tool that auto-generate one for me so far ^^
So if you can give me a finished ascii art, I'd put it in there as a comment

**Comment by RoiEXLab** [comment]
@ssoloff If you can provide some ASCII art to clarify this test, I'd like to see it in a follow-up PR because I won't be changing this until tomorrow.
All of your comments should be addressed.

**Comment by RoiEXLab** [comment]
@ron-murhammer Merging this now.
I have a follow-up PR ready to open that increases the performance by a noticeable amount.
